### PR TITLE
Implement hash_join for merges

### PR DESCRIPTION
**Pull request by phofl**
cc @mroeschke 

Our abstraction in merges is bad, this makes it a little worse unfortunately. But it enables a potentially huge performance improvement for joins that could be hash joins. I am using "right" to make the decision, because left determines the result order, which means that we would have to sort after we are finished which gives the performance improvement back. using right makes this problem go away. 

We get time complexity O(m+n) here over O(m*n) with a non-trivial factor as before

This makes adding semi joins pretty easy as well, which is nice next to the performance improvements here.

```
| Change   | Before [38086f11] <backtest>   | After [3b6b787e] <to>   |   Ratio | Benchmark (Parameter)                                                       |
|----------|--------------------------------|-------------------------|---------|-----------------------------------------------------------------------------|
| -        | 193±4ms                        | 165±5ms                 |    0.85 | join_merge.I8Merge.time_i8merge('inner')                                    |
| -        | 7.76±0.2ms                     | 6.53±0.06ms             |    0.84 | join_merge.Merge.time_merge_2intkey(False)                                  |
| -        | 1.03±0.02ms                    | 834±4μs                 |    0.81 | join_merge.Merge.time_merge_dataframe_integer_2key(False)                   |
| -        | 485±2μs                        | 339±5μs                 |    0.7  | join_merge.Merge.time_merge_dataframe_integer_key(False)                    |
| -        | 3.29±0.2ms                     | 2.29±0.07ms             |    0.7  | join_merge.MergeDatetime.time_merge(('ms', 'ms'), 'Europe/Brussels', False) |
| -        | 3.31±0.07ms                    | 2.27±0.07ms             |    0.69 | join_merge.MergeDatetime.time_merge(('ms', 'ms'), None, False)              |
| -        | 2.79±0.09ms                    | 1.77±0.01ms             |    0.63 | join_merge.MergeDatetime.time_merge(('ns', 'ms'), 'Europe/Brussels', False) |
| -        | 2.89±0.04ms                    | 1.78±0.05ms             |    0.62 | join_merge.MergeDatetime.time_merge(('ns', 'ms'), None, False)              |
| -        | 2.57±0.09ms                    | 1.56±0.03ms             |    0.61 | join_merge.MergeDatetime.time_merge(('ns', 'ns'), None, False)              |
| -        | 1.97±0.05ms                    | 1.18±0.02ms             |    0.6  | join_merge.MergeEA.time_merge('Float32', False)                             |
| -        | 1.84±0.02ms                    | 1.10±0.03ms             |    0.6  | join_merge.MergeEA.time_merge('UInt16', False)                              |
| -        | 2.09±0.04ms                    | 1.24±0.01ms             |    0.59 | join_merge.MergeEA.time_merge('UInt64', False)                              |
| -        | 2.10±0.09ms                    | 1.22±0.01ms             |    0.58 | join_merge.MergeEA.time_merge('Float64', False)                             |
| -        | 2.09±0.08ms                    | 1.22±0.01ms             |    0.58 | join_merge.MergeEA.time_merge('UInt32', False)                              |
| -        | 2.70±0.1ms                     | 1.54±0.02ms             |    0.57 | join_merge.MergeDatetime.time_merge(('ns', 'ns'), 'Europe/Brussels', False) |
| -        | 1.72±0.02ms                    | 971±20μs                |    0.57 | join_merge.MergeEA.time_merge('Int16', False)                               |
| -        | 1.76±0.03ms                    | 973±10μs                |    0.55 | join_merge.MergeEA.time_merge('Int32', False)                               |
| -        | 1.94±0.09ms                    | 1.07±0.03ms             |    0.55 | join_merge.MergeEA.time_merge('Int64', False)                               |
| -        | 57.0±2ms                       | 21.7±0.3ms              |    0.38 | join_merge.UniqueMerge.time_unique_merge(1000000)                           |
| -        | 106±7ms                        | 34.1±0.3ms              |    0.32 | join_merge.UniqueMerge.time_unique_merge(4000000)                           |
```

### CODE REVIEW

File: pandas/_libs/hashtable_class_helper.pxi.in [hunk 1]
```diff
@@ -545,6 +545,38 @@ cdef class {{name}}HashTable(HashTable):
                     self.table.vals[k] = i
         self.na_position = na_position
 
+    @cython.boundscheck(False)
```
Comment (WillAyd) [inline pandas/_libs/hashtable_class_helper.pxi.in:548]: Probably worth adding `@cython.wraparound(False) here as well

File: pandas/_libs/hashtable_class_helper.pxi.in [hunk 2]
```diff
@@ -545,6 +545,38 @@ cdef class {{name}}HashTable(HashTable):
...
+    def hash_inner_join(self, const {{dtype}}_t[:] values, const uint8_t[:] mask = None) -> tuple[ndarray, ndarray]:
+        cdef:
+            Py_ssize_t i, n = len(values)
+            int ret = 0, ctr = 0
+            {{c_type}} val
+            khiter_t k
+            intp_t[::1] locs = np.empty(n, dtype=np.intp)
```
Comment (WillAyd) [inline pandas/_libs/hashtable_class_helper.pxi.in:555]: The Vector classes would be a better fit than the allocate-in-full + trim strategy

File: pandas/core/reshape/merge.py [hunk 3]
```diff
@@ -2527,28 +2535,41 @@ def _factorize_keys(
...
     else:
         # Argument 1 to "factorize" of "ObjectFactorizer" has incompatible type
         # "Union[ndarray[Any, dtype[signedinteger[_64Bit]]],
         # ndarray[Any, dtype[object_]]]"; expected "ndarray[Any, dtype[object_]]"
-        llab = rizer.factorize(lk)  # type: ignore[arg-type]
-        rlab = rizer.factorize(rk)  # type: ignore[arg-type]
+        lk_data, rk_data = lk, rk  # type: ignore[assignment]
+        lk_mask, rk_mask = None, None
+
+    hash_join_available = hash_join_available and lk.dtype.kind in "iufb"
```
Comment (WillAyd) [inline pandas/core/reshape/merge.py:2561]: The determination of whether or not to use a hash join looks split across a few different functions - is there any path to better consolidating that?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/reshape/merge.py (hunk 3 above)
Comment (phofl) [reply pandas/core/reshape/merge.py:2561]: We can push ``how`` into this method as well, that would consolidate. Good call. I had this at a different place initially, but this created overhead in cases where we couldn't use it, that's why it ended here.

File: pandas/_libs/hashtable_class_helper.pxi.in (hunk 1 above)
Comment (phofl) [reply pandas/_libs/hashtable_class_helper.pxi.in:548]: Thx

File: pandas/_libs/hashtable_class_helper.pxi.in (hunk 2 above)
Comment (phofl) [reply pandas/_libs/hashtable_class_helper.pxi.in:555]: Oh, I didn't think about them. You are definitely correct, thx. Changed

File: pandas/_libs/hashtable_class_helper.pxi.in [hunk 4]
```diff
@@ -557,6 +557,49 @@ cdef class {{name}}HashTable(HashTable):
...
+        if self.uses_mask and mask is None:
+            raise NotImplementedError  # pragma: no cover
+
+        with nogil:
+            for i in range(n):
+                if self.uses_mask and mask[i]:
+                    if self.na_position == -1:
+                        continue
+                    if needs_resize(l.size, l.capacity):
+                        with gil:
```
Comment (WillAyd) [inline pandas/_libs/hashtable_class_helper.pxi.in:585]: You shouldn't need to reacquire the gil here - was Cython giving an error?

**Review by WillAyd — COMMENTED** [review]
Thanks for the refactor - this structure makes a lot more sense

File: pandas/_libs/hashtable_class_helper.pxi.in (hunk 4 above)
Comment (phofl) [reply pandas/_libs/hashtable_class_helper.pxi.in:585]: Yep

```
Discarding owned Python object not allowed without gil
```

Comment (WillAyd) [reply pandas/_libs/hashtable_class_helper.pxi.in:585]: Huh OK - I see that pattern elsewhere in the codebase. I think there is something simple being overlooked that requires the GIL, but let's leave that to a separate PR

**Review by WillAyd — APPROVED** [review]
nice work @phofl

File: pandas/_libs/hashtable_class_helper.pxi.in (hunk 4 above)
Comment (phofl) [reply pandas/_libs/hashtable_class_helper.pxi.in:585]: > I think there is something simple being overlooked that requires the GIL

That would make me very happy as a Dask developer :)

**Comment by phofl** [comment]
Thx for your rewiew! merging to unblock follow up work but happy to address other comments in follow ups

**Pull request merged**